Nucleotide Diversity

philipschulz opened the pull request:
- Nucleotide Diversity chart component for Focus page
- Nucleotide Diversity multi chart component for CompareEquals page

philipschulz committed:
data fetch setup

philipschulz committed:
calc nucleotide diversity

philipschulz committed:
calculate mean entropy of the sample

philipschulz committed:
NaN log fix

philipschulz committed:
barchart skeleton

philipschulz committed:
position shift

philipschulz committed:
custom tooltip

philipschulz committed:
plot controls

philipschulz committed:
mutation proportions in tooltip

philipschulz committed:
tooltip properties

philipschulz committed:
prettier formatting

philipschulz committed:
get weekly mutation proportions

philipschulz committed:
basic entropy over time plot

philipschulz committed:
decode nuc and aa mutations

philipschulz committed:
move mutationQuery into useData

philipschulz committed:
reorder entropy calculation

philipschulz committed:
variable brush index lookup

philipschulz committed:
fixed nuc aa switch error

philipschulz committed:
all combinations of plot types and nuc/aa working

philipschulz committed:
AA entropy sorted and labeled

philipschulz committed:
AA brush index lookup, fixed AA positioning

philipschulz committed:
cleanup

philipschulz committed:
data transformation, plot styling

philipschulz committed:
fixed loader

philipschulz committed:
fixed brush index bug

philipschulz committed:
fixed gene select zoom, added multi gene select

philipschulz committed:
entropy over time gene multiselect

philipschulz committed:
gene multiselect working for AA

philipschulz committed:
gene colors

philipschulz committed:
gene label colors

philipschulz committed:
added multi plot skeleton to CompareEqualsPage

philipschulz committed:
semi-working multi lineage entropy plot

philipschulz committed:
rename nucleotide diversity to nucleotide entropy

philipschulz committed:
correct grouping of entropy values by day

philipschulz committed:
fix no variant crash, calculate daterange directly from selector

philipschulz committed:
remove basicVariantDataQuery

philipschulz committed:
nice x axis date formatting

philipschulz committed:
main plot tick formatting, added option to include deletion mutations in entropy calculation

philipschulz committed:
fixed the multi variant plot display bug?

philipschulz committed:
display bug

philipschulz committed:
toggle to show positions with 0 entropy & colored gene regions

philipschulz committed:
fixed color assignment

philipschulz committed:
text changes

philipschulz committed:
comments

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+    selectors: LapisSelector[];
+  };
+
+type Variant = {
```
Do we need this type? According to my IDE it is unused.

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
eslint (`npx eslint .`) complains about a couple of lines here and in `NucleotideEntropy.tsx`. Could you please fix those?

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+            <PipeDividedOptionsButtons
+              options={[
+                { label: 'NUC', value: 'nuc' },
+                { label: 'AA', value: 'aa' },
```
The surrounding diagrams use the full names.
```suggestion
                { label: 'Nucleotides', value: 'nuc' },
                { label: 'Amino acids', value: 'aa' },
```

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
Please use prettier to format the files. If you use IntelliJ, there is a nice plugin :)

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+export const NucleotideEntropyMultiChart = ({ selectors }: Props) => {
+    const [sequenceType, setSequenceType] = useState<SequenceType>('nuc');
+    const [gene, setGene] = useState<string>('All');
+    const [deletions, setDeletions] = useState<boolean>(false);
```
```suggestion
    const [includeDeletions, setIncludeDeletions] = useState<boolean>(false);
```
That would make it more obvious that the variable is a boolean.

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+
+
+    let plotArea;
+    if (variants![0].length == 0) {
```
That `!` looks supsicious to me. This basically disables typesafety. Why is the type nullable if you seem to be sure that it is not null here?

fengelniederhammer replied:
Could you please double check all the `!` in your code? This potentially results in errors.

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+
+    const controls = (
+        <div className='mb-4'>
+          {/* AA vs. nucs */}
```
Do we need those comments here? I'd propose to
* either simply delete them
* or extract the code that the comment refers to into a separate component and give that component a name that replaces the comment.

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+              />
+            </FormGroup>
+          {/* Genes */}
+          {
```
I think we don't need this pair of brackets, do we?

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+              <Form.Control
+                as='select'
+                value={gene}
+                onChange={ev => setGene(ev.target.value)}
```
What is `ev`? Could you please give it a proper name? :)

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+  const weeklyVariantMutationProportionQuery = useQuery(
+    async signal => {
+      //calculate weeks
+      const dayArray: UnifiedDay[] = [selectors[0].dateRange?.getDateRange().dateFrom!, selectors[0].dateRange?.getDateRange().dateTo!];
```
Explicit type declarations are usually not necessary. Typescript is smart enough to figure out the type by itself.
```suggestion
      const days = [selectors[0].dateRange?.getDateRange().dateFrom!, selectors[0].dateRange?.getDateRange().dateTo!];
```

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+      for (let i = 0; i < weeks.length; i++) {
+        let dateRange: DateRange = { dateFrom: weeks[i].firstDay, dateTo: weeks[i].firstDay };
+        weekDateRanges[i] = dateRange;
+      }
```
Using `Array.map` is usually more idiomatic are easier to read:
```suggestion
      const weekDateRanges: DateRange[] = globalDateCache
        .weeksFromRange({
          min: dayRange.min.isoWeek,
          max: dayRange.max.isoWeek,
        })
        .map(week => ({
          dateFrom: week.firstDay,
          dateTo: week.firstDay,
        }));
```

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+    [selectors, sequenceType]
+  );
+    
+  const data = useMemo(() => {
```
This could be inlined:
```suggestion
  return useMemo(() => {
```

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+            dateRange: new FixedDateRangeSelector(w),
+          })
+        };
+      })
```
Same here: (Leveraging ChatGPT :) )
```suggestion
        const weekSelectors = selectors.flatMap(selector =>
            weekDateRanges.map(w => ({
                ...selector,
                dateRange: new FixedDateRangeSelector(w),
            })),
        );
```

fengelniederhammer replied:
This is the explanation that ChatGPT gave:
This version leverages the flatMap function to flatten the resulting nested arrays into a single-level array while applying the provided mapping function. In this case, for each selector element, we create a new array of LapisSelector objects with the updated dateRange property, and then flatten the resulting arrays into one continuous array.

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+        sequenceType: sequenceType,
+        weekNumber: weekLength,
+        result: await Promise.all(
+          weekSelectors.map((w, i) =>
```
Let's also use proper variable names here, please: 
```suggestion
          weekSelectors.map((weekSelector, i) =>
```

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+        result: await Promise.all(
+          weekSelectors.map((w, i) =>
+            MutationProportionData.fromApi(w, sequenceType, signal).then(data => {
+              const proportions: MutationProportionEntry[] = data.payload.map(m => m);
```
Why do we need this line? It essentially does nothing.

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+    deletions: boolean
+): any | undefined => {
+  //fetch the proportions per position in weekly segments
+  const weeklyVariantMutationProportionQuery = useQuery(
```
Let's also get rid of the comments here: :) 
```suggestion
  const weeklyVariantMutationProportions = useProportionsPerPositionInWeeklySegments(selectors, sequenceType);

```

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+      return undefined;
+    }
+    let sequenceType = weeklyVariantMutationProportionQuery.data.sequenceType;
+    if (!sequenceType) {
```
Can `sequenceType` ever be falsy?

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+      let weekLength = weekDateRanges.length;
+
+      return {
+        sequenceType: sequenceType,
```
Why do you return `sequenceType` here? This is the same as the input.

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+      return undefined;
+    }
+    if (variants.length*weekNumber != weeklyVariantMutationProportionQuery.data.result.length){
+      return undefined;
```
Why return undefined in this case? Isn't this rather an error? Is it important to check this here?

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+        dayMap.set(day, tt);
+      }
+    });
+    const dayArray = Array.from(dayMap.values()).slice(0, -1); //depending on the day, the latest week just started, so the entropy is calculated as 0 because there are no samples
```
Also according to ChatGPT, this can be simplified to this:
```suggestion
const dayMap = timeData.reduce((map, tt) => {
  const day = tt.day;
  const dayGroup = map[day] || {};
  return {
    ...map,
    [day]: { ...dayGroup, ...tt }
  };
}, {});

const dayArray = Object.values(dayMap).slice(0, -1);
```

Maybe we can also improve the naming of variables here? `tt` should be something understandable, also including `Map` or `Array` usually has no benefit in readability. `Map` can usually be replaced by something like `daysGroupedBy<keys>`. Instead of `Array`, you can simply use the plural (here `days`).

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+
+const Plot = ({ plotData, ticks, variants, sequenceType }: PlotProps) => {
+return (
+    <>
```
I think we don't need this fragment, do we?

fengelniederhammer commented on src/components/NucleotideEntropy.tsx:
```diff
@@ -0,0 +1,706 @@
+
+
+export const NucleotideEntropy = ({ selector }: Props) => {
+  const [plotType, setPlotType] = useState<string>('time');
```
I think we can improve the typing (and naming) here:
```suggestion
  const [plotType, setPlotType] = useState<'time' | 'position'>('time');
```

fengelniederhammer commented on src/components/NucleotideEntropy.tsx:
```diff
@@ -0,0 +1,706 @@
+  }
+
+  return (
+    <>
```
This also looks like we don't need that fragment.

fengelniederhammer commented on src/components/NucleotideEntropy.tsx:
```diff
@@ -0,0 +1,706 @@
+  const controls = (
+    <div className='mb-4'>
+      {/* AA vs. nucs */}
+      <div className='mb-2 flex'>
```
I would guess that we could extract some of those components to reuse them here and in `NucleotideEntropyMultiChart`

fengelniederhammer commented on src/components/NucleotideEntropy.tsx:
```diff
@@ -0,0 +1,706 @@
+      ? posEntropy
+      : posEntropy.filter(g => g.position.includes(gene.value));
+  const sum = filteredPos.map(f => f.entropy).reduce((x, a) => x + a, 0);
+  const count =
+    sequenceType === 'nuc'
+      ? gene.endPosition - gene.startPosition
+      : (gene.value === 'All' ? jsonRefData.genes : jsonRefData.genes.filter(g => g.name.includes(gene.value)))
+          .map(g => g.aaSeq.length)
+          .reduce((x, a) => x + a, 0);
```
This is very hard to read. Could you maybe extract parts of it?

fengelniederhammer commented on src/components/NucleotideEntropy.tsx:
```diff
@@ -0,0 +1,706 @@
+  }
+};
+
+const CalculateEntropy = (
```
This should have a name starting with lower case. Only React components should start with a capital letter.

fengelniederhammer commented on src/components/NucleotideEntropy.tsx:
```diff
@@ -0,0 +1,706 @@
+
+
+  //convert proportion to entropy
+  positionGroups.map(p => {
```
Either use `forEach` or return something from the array function in `map`, please.

fengelniederhammer reviewed (changes requested):
Please check the code for unnecessary things: There are a couple of fragments, variables and other things that we could get rid of.

Also I would propose to restructure the code:
* create a folder `/src/components/NucleotideEntropy`
* In that folder, create an `index.tsx` that exposes `NucleotideEntropy` and `NucleotideEntropyMultiChart` (either by reexport or by holding the actual implementation). That way, consumers only ever need to know that index.tsx and no other file in the folder.
* Hooks go into separate files in that folder.
* Also extract components into separate files: Either if they are large by themselves, or if they can be reused by multiple other components.
* In general methods should not be longer than 100 lines. Files should also not be much longer than that either.

Also please add unit tests to make sure that your code actually works. In the ideal case, you can split your tests into display and logic tests:
* test the hooks separately. This is where the logic sits. Make sure that the computation is correct. "Given the server returns some specific data, then the hooks computes xyz...".
* test the components separately. Mock the hooks. "Given some specific data to display, then the component shows..."

Also please try to get rid of most of the comments in the code. Comments tend to be outdated and get misplaced, because people forget to update them when changing the code. Some few comments might be justified, but usually one can apply any of:
* The code is already expressive enough and the comment only paraphrases what's written in the code. In that case, simply delete the comment.
* The code explains what's going on in the code. It also paraphrases the code, but the code is more complex, so that it deserves some explanation. In that case, extract the code into a class/method/component (whatever fits) and give it a name such that you don't need the comment, because it's included in the name.

And please use expressive names for all variables. It really helps other people to understand the code a lot faster, if names communicate the purpose of variables.

fengelniederhammer commented on src/components/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,294 @@
+  genes.push({ name: 'All', startPosition: 0, endPosition: 29903, aaSeq: ''});
+
+function nonNullValues(obj: any) {
+  return Object.entries(obj).filter(([key, value]) => value !== undefined).map(([key, value]) => value).toString();
```
Calling unused variables `_` is a common pattern, if you somehow have to include them, but you don't actually need them:
```suggestion
  return Object.entries(obj).filter(([_, value]) => value !== undefined).map(([_, value]) => value).toString();
```

fengelniederhammer commented:
You might also want to rebase your branch to the current develop branch. It's quite a few commits behind. :)

philipschulz committed:
cleanup

philipschulz committed:
restructured NucelotideEntropy, plot utilities and entropy calc into seperate components

philipschulz committed:
prettier format FocusCompareEqualsPage

fengelniederhammer committed:
refactor: improve naming

fengelniederhammer committed:
refactor: split Plot component

fengelniederhammer committed:
refactor: extract controls components

fengelniederhammer committed:
refactor: extract hook to separate file

fengelniederhammer committed:
test: test nucleotide entropy data hook

fengelniederhammer committed:
refactor: split hook into smaller parts

fengelniederhammer committed:
refactor: functional style instead of loops

fengelniederhammer committed:
refactor: split nucleotide entropy hook into two separate hooks

fengelniederhammer committed:
test: also test including zero entropy positions

fengelniederhammer committed:
refactor: NucleotideEntropyMultiChart

fengelniederhammer committed:
refactor: more functional style

fengelniederhammer committed:
fix: workaround because custom useQuery doesn't immediately update the data

This is a workaround for a weird bug: When changing to amino acids, the plot crashes.
On the first update run, the old data for nucleotides would still be returned by useQuery.
Only on the second run, it returns the new data.

Also make the unit test independent of local time.

fengelniederhammer committed:
refactor: extract hook to separate file

fengelniederhammer commented on src/components/NucleotideEntropy/NucleotideEntropyMultiChart.tsx:
```diff
@@ -0,0 +1,196 @@
+  };
+
+  const exploreUrl = useExploreUrl()!;
+  const variants = (exploreUrl.variants ?? []).map(nonNullValuesToString);
```
@chaoran-chen Does this particular line make sense? I'm not so sure about it. It seems to work, but it might be quite fragile.

fengelniederhammer committed:
fix: use date range for whole week instead of only for the first day

fengelniederhammer reviewed (approved):
From my side it looks good now, but it would be good if someone else could double check my refactorings ;)

chaoran-chen reviewed (changes requested):
By default, the positions with zero entropy should be included. In fact, I think that we should negate the checkbox to "Exclude positions with zero entropy"

![image](https://github.com/GenSpectrum/cov-spectrum-website/assets/18666552/e06bc897-455a-402f-918f-10623f650242)

Unfortunately, I see that the plot has some performance issues when the positions with zero entropy are included: It takes a long time to load and the zooming is very slow. @fengelniederhammer , do you see a way to improve the performance?

fengelniederhammer committed:
refactor: invert checkbox so that it now reads "exclude positions with zero entropy"

fengelniederhammer committed:
refactor: don't unnecessarily call the function twice

fengelniederhammer commented:
I'm not sure whether we can do something here. It think it just takes the browser very long to render 30.000 bars. 

The computation is clearly not the problem: When changing the threshold from 0.1 to 0, the data is the same (cached by useMemo). The browser just has to rerender the plot.

Why is it actually useful to see mostly empty bars?

chaoran-chen commented:
If we exclude the zero entropy bars, the x-axis is not "uniform" anymore. It jumps from 2 to 12, etc.

![image](https://github.com/GenSpectrum/cov-spectrum-website/assets/18666552/a80402d1-d2a6-46f8-9cf5-e9d487fdd684)

Would it be possible to not actually render the zero entropy positions but just leave the space empty?

chaoran-chen reviewed (approved):
As further performance improvements seem difficult at the moment, I'll merge it.

Thank you very much, @philipschulz and @fengelniederhammer for the great work!
